test/generate/build_test.dart

```diff
@@ -0,0 +1,224 @@
+          'a|lib/b.txt': 'b',
+        }, outputs: {
+          'a|web/a.txt.copy': 'a',
+          'a|web/a.txt.clone': 'a',
+          'a|lib/b.txt.copy': 'b',
+          'a|lib/b.txt.clone': 'b',
+          'a|web/a.txt.clone.copy.0': 'a',
+          'a|web/a.txt.clone.copy.1': 'a',
+        });
+      });
+    });
+
+    group('inputs from other packages', () {
+      test('only gets inputs from lib, can output to root package', () async {
+        var phases = [
+          [
+            new Phase(
+                [new CopyBuilder(outputPackage: 'a')], [new InputSet('b')]),
+          ]
+        ];
+        await testPhases(phases, {'b|web/b.txt': 'b', 'b|lib/b.txt': 'b',},
+            outputs: {'a|lib/b.txt.copy': 'b',});
+      });
+
+      test('can\'t output files in non-root packages', () async {
+        var phases = [
+          [
+            new Phase([new CopyBuilder()], [new InputSet('b')]),
+          ]
+        ];
+        await testPhases(phases, {'b|lib/b.txt': 'b',},
+            outputs: {}, status: BuildStatus.Failure);
+      },
+          skip: 'Failing, InMemoryAssetWriter doesn\'t throw. Need to handle '
+              'this in BuildStep instead of AssetWriter.');
+    });
+
+    test('multiple phases, inputs from multiple packages', () async {
+      var phases = [
+        [
+          new Phase([new CopyBuilder()], [new InputSet('a')]),
+          new Phase([
+            new CopyBuilder(extension: 'clone', outputPackage: 'a')
+          ], [
+            new InputSet('b', filePatterns: ['**/*.txt']),
+          ]),
+        ],
+        [
+          new Phase([
+            new CopyBuilder(numCopies: 2, outputPackage: 'a')
+          ], [
+            new InputSet('a', filePatterns: ['lib/*.txt.*']),
+            new InputSet('b', filePatterns: ['**/*.dart']),
+          ]),
+        ]
+      ];
+      await testPhases(phases, {
+        'a|web/1.txt': '1',
+        'a|lib/2.txt': '2',
+        'b|lib/3.txt': '3',
+        'b|lib/b.dart': 'main() {}',
+      }, outputs: {
+        'a|web/1.txt.copy': '1',
+        'a|lib/2.txt.copy': '2',
+        'a|lib/3.txt.clone': '3',
+        'a|lib/2.txt.copy.copy.0': '2',
+        'a|lib/2.txt.copy.copy.1': '2',
+        'a|lib/3.txt.clone.copy.0': '3',
+        'a|lib/3.txt.clone.copy.1': '3',
+        'a|lib/b.dart.copy.0': 'main() {}',
+        'a|lib/b.dart.copy.1': 'main() {}',
+      });
+    });
+  });
+
+  group('errors', () {
+    test('when overwriting files', () async {
+      var phases = [
+        [
+          new Phase([new CopyBuilder()], [new InputSet('a')]),
+        ]
+      ];
+      await testPhases(phases, {'a|lib/a.txt': 'a', 'a|lib/a.txt.copy': 'a'},
+          status: BuildStatus.Failure,
+          exceptionMatcher: invalidOutputException);
+    },
+        skip: 'InMemoryAssetWriter doesn\'t check this, should be handled at '
+            'the BuildStep level.');
+  });
+
+  test('outputs from builds shouldn\'t be inputs to later ones', () {},
```

Comment:
I will rename the tests, its for back to back full builds. Files created by one full build shouldn't be treated as inputs for other full builds.